src/main/java/com/conveyal/analysis/components/TaskScheduler.java

```diff
@@ -143,19 +147,34 @@ public final void run () {
         }
     }
 
-    ////// Methods for reporting active tasks for each user
-
-    /** Return an empty list even when no tasks have been recorded for the user (return is always non-null). */
-    public List<Task> getTasksForUser (String userEmail) {
-        Set<Task> tasks = tasksForUser.get(userEmail);
-        return tasks == null ? Collections.emptyList() : List.copyOf(tasks);
+    /**
+     * Return the status of all background tasks for the given user, as API model objects for serialization to JSON.
+     * Completed tasks that finished over a minute ago will be purged after returning them. This ensures they're sent
+     * at least once to the UI and gives any other tabs a chance to poll for them.
+     * Conversion to the API model is done here to allow synchronization without copying the list of internal tasks.
+     * The task scheduler collections are being updated by worker threads. The fields of the individual Tasks may also
+     * be updated at any time. So there is a risk of sending a partially updated Task out to the UI. If this ever causes
+     * problems we'll need to lock each Task independently.
+     * Returns an empty list even when no tasks have been recorded for the user (return is always non-null).
+     */
+    public @Nonnull List<ApiTask> getTasksForUser (String userEmail) {
+        synchronized (tasksForUser) {
+            Set<Task> tasks = tasksForUser.get(userEmail);
+            if (tasks == null) return Collections.emptyList();
+            List<ApiTask> apiTaskList = tasks.stream()
+                    .map(Task::toApiTask)
```

Comment:
Should this method live in the ApiTask? `ApiTask::fromTask`.

Reply:
I tried it both ways. Both have shortcomings. Ideally classes for the API would not be accessing non-final fields of internal data structures, and the internal data structures are in packages that ideally shouldn't even know the API exists. Respecting both constraints yields lots of verbose Java repetition and boilerplate. The real solution is language support for properties and immutable data classes (port to Swift)?

Reply:
Hmmm. Putting getters on the non-final fields of `Task` certainly seems more ideal than having `BackgroundTask` import a subclass from a Controller. What happens if we want to have another type of `ApiTask`? For example, for admin purposes?